sdk/eventhubs/azure_messaging_eventhubs/src/event_processor/processor.rs

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) Microsoft Corp. All Rights Reserved.
+// Licensed under the MIT License.
+
+//use async_channel::{bounded, Receiver, Sender};
+use super::{
+    load_balancer::LoadBalancer,
+    models::{Checkpoint, StartPositions},
+    partition_client::PartitionClient,
+    CheckpointStore, ProcessorStrategy,
+};
+use crate::{
+    models::{ConsumerClientDetails, EventHubProperties},
+    ConsumerClient, OpenReceiverOptions, StartLocation, StartPosition,
+};
+//use async_io::Timer;
+use async_lock::Mutex as AsyncMutex;
+use azure_core::{error::ErrorKind as AzureErrorKind, Error, Result};
+use futures::{
+    channel::mpsc::{channel, Receiver, Sender},
+    SinkExt, StreamExt,
+};
+use std::{
+    sync::{
+        Arc,
+        Mutex as SyncMutex, // Mutex for blocking operations
+    },
+    time::Duration,
+    {collections::HashMap, sync::Weak},
+};
+use tracing::{debug, error, info};
+
+/// Represents the event processor responsible for processing events
+/// from Event Hub partitions.
+///
+/// This struct manages the load balancing strategy, checkpoint store,
+/// and consumer client for processing events.
+/// It provides methods for starting the event processor, dispatching
+/// events, and managing partition clients.
+///
+/// The event processor uses a load balancer to distribute the load
+/// across partitions and a checkpoint store to manage checkpoints.
+///
+/// For more information on Event Processors and scenarios in which you would
+/// use an Event Processor, see the [Event Processor documentation](https://learn.microsoft.com/azure/event-hubs/event-processor-balance-partition-load).
+///
+pub struct EventProcessor {
+    checkpoint_store: Arc<dyn CheckpointStore + Send + Sync>,
+    load_balancer: Arc<AsyncMutex<LoadBalancer>>,
+    consumer_client: Arc<ConsumerClient>,
+    next_partition_clients: AsyncMutex<Receiver<Arc<PartitionClient>>>,
+    next_partition_client_sender: Sender<Arc<PartitionClient>>,
+    client_details: ConsumerClientDetails,
+    prefetch: u32,
+    update_interval: Duration,
+    start_positions: StartPositions,
+    is_running: std::sync::Mutex<bool>,
```

Comment:
Any reason not to use an `AsyncMutex` here? At first glance, seems odd to mix them.

Reply:
Because there is no case where I need to do an asynchronous operation while holding the lock. This lock simply protects the single "bool" value and all I ever do is check it's value or set it to false.

I use an `AsyncMutex` when I need a mutex that can span across asynchronous calls.